### ISSUE TITLE
state_quant returns float32 for float64 inputs

### ISSUE THREAD

**Issue by GraceDyer** [jeshraghian/snntorch#439]
**Issue type**

- [x] Bug Report
- [ ] Feature Request
- [ ] Help wanted
- [ ] Other

**snnTorch version**

`1.0.0`

**Description**

I found a reproducible dtype bug in the public helper `snntorch.functional.quant.state_quant`.

When the input tensor uses `torch.float64`, the quantized output is returned as `torch.float32`. This occurs with both uniform and non-uniform quantization.

As a result, applying `state_quant` to a double-precision hidden state silently changes its dtype. This can cause precision loss or dtype mismatches in a model converted with `.double()`.

From source inspection, the quantization levels are created using `torch.linspace(...)` or `torch.tensor(...)` without specifying a dtype. In `StateQuant.forward`, the levels are moved to the input device with `levels.to(device)`, but they are not converted to the input dtype. The returned tensor is then selected directly from those levels, so it inherits their default `float32` dtype.

I reproduced this with both snnTorch `1.0.0` and the current master commit in fresh CPU-only Docker containers using Debian GNU/Linux 12, Python 3.11.16, and PyTorch 2.8.0+cpu.

**Minimal code to reproduce the error/bug**

```python
import torch
from snntorch.functional import quant

torch.set_default_dtype(torch.float32)

x = torch.tensor([0.125, 0.875], dtype=torch.float64)

uniform_quant = quant.state_quant(
    num_bits=4,
    uniform=True,
    threshold=1.0,
)
nonuniform_quant = quant.state_quant(
    num_bits=4,
    uniform=False,
    thr_centered=False,
    threshold=1.0,
)

uniform_output = uniform_quant(x)
nonuniform_output = nonuniform_quant(x)

print("input_dtype =", x.dtype)
print("uniform_output_dtype =", uniform_output.dtype)
print("nonuniform_output_dtype =", nonuniform_output.dtype)
```

**What is expected**

`state_quant` should preserve the floating-point dtype of the input tensor.

For a `torch.float64` input, both quantized outputs should also use `torch.float64`.

**What actually happens**

Both uniform and non-uniform quantization return `torch.float32` tensors for a `torch.float64` input.

Example output from the Docker reproduction:

```text
input_dtype = torch.float64
uniform_output_dtype = torch.float32
nonuniform_output_dtype = torch.float32
```

**Comment by ixfd64** [comment]
Should be resolved in #444. Let us know if there are still issues.
